### PR TITLE
BUG: reshape fix for maybe_infer_to_datetimelike()

### LINKED ISSUE #16362: BUG: Transpose construction/block error with certain dtypes
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
from io import StringIO

data = "dt1,obj1,obj2,obj3,num1\n,a,c,e,0\n,b,d,f,1"

df1 = pd.read_csv(StringIO(data), dtype={'num1': float}, parse_dates=['dt1'])
df2 = pd.read_csv(StringIO(data), dtype={'num1': int}, parse_dates=['dt1'])

print(df1.T)
print(df2.T)  # ValueError in v0.20.1 but not v0.19.2

```
#### Problem description

Tried to provide a minimal example here. I originally encountered the error with a dataframe containing datetime, object, and int dtypes, and triggered it with this...

`df.iloc[[0, -1]].T`

...and even found things like this:

```
df.iloc[[0, 1], [2, 3, 4, 9]].T  # okay
df.iloc[[0, 1], [2, 3, 4, 4, 9]].T  # ValueError
df.iloc[[0, 1], [2, 3, 4, 9, 4]].T  # okay (?!)
```

#### Expected Output
        0    1
dt1   NaT  NaT
obj1    a    b
obj2    c    d
obj3    e    f
num1    0    1
        0    1
dt1   NaT  NaT
obj1    a    b
obj2    c    d
obj3    e    f
num1    0    1

#### Output of ``pd.show_versions()``
<details>
# Paste the output here pd.show_versions() here
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.0.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 61 Stepping 4, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.20.1
pytest: 3.0.5
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
xarray: None
IPython: 5.1.0
sphinx: 1.5.1
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2016.10
blosc: None
bottleneck: 1.2.0
tables: 3.2.2
numexpr: 2.6.2
feather: None
matplotlib: 2.0.1
openpyxl: 2.4.1
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 3.7.2
bs4: 4.5.3
html5lib: None
sqlalchemy: 1.1.5
pymysql: None
psycopg2: None
jinja2: 2.9.4
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by RobinFiveWords**
- [x] closes #16362
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [ ] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
tests should *always* be included. generally I write them *first* to ensure that they are picking up the bug. you can push to this PR itself.

**Comment by RobinFiveWords** [comment]
@jreback I might need some guidance...I might not have needed to touch this -- the branch had a new conflict, I imagine because of other merged PRs, but only with whatsnew. Now I'm not sure how to fix it back to a state where tests pass.

File: pandas/tests/dtypes/test_cast.py [hunk 1]
```diff
@@ -213,6 +213,13 @@ def test_maybe_convert_scalar(self):
         result = maybe_convert_scalar(Timedelta('1 day 1 min'))
         assert result == Timedelta('1 day 1 min').value
 
+    def test_maybe_infer_to_datetimelike(self):
+
```
Comment (jreback) [inline pandas/tests/dtypes/test_cast.py]: add the issue number here

File: pandas/tests/dtypes/test_cast.py [hunk 2]
```diff
@@ -213,6 +213,13 @@ def test_maybe_convert_scalar(self):
...
+        # this construction raises IndexError in pandas=0.20.1
+        result = DataFrame(np.array([[NaT, 'a', 'b', 0],
```
Comment (jreback) [inline pandas/tests/dtypes/test_cast.py]: use both examples from here:  https://github.com/pandas-dev/pandas/issues/16362#issuecomment-301581166

and assert the result to an expected result == 8

**Review by jreback — CHANGES_REQUESTED** [review]
rebase on master

```
git pull origin/master
git rebase -i origin/master
```

then when pushing
```
git push yourupstream yourbranch -f
```

**Comment by jreback** [comment]
pls add a whatsnew note in 0.20.2 (bug fixes, reshaping section)

**Comment by TomAugspurger** [comment]
Does https://github.com/pandas-dev/pandas/commit/cd11b497608d216dbd4276ddc119f52dc0588a70 look like the full set of changes, aside from the whatsnew?

**Comment by TomAugspurger** [comment]
It looks like you lost some changes, but you should be able to recover them with your git reflog, assuming you haven't cleared it manually. Or maybe even a `git reset --soft cd11b49`. If you ping me on gitter I can walk you through the recover steps.

**Comment by RobinFiveWords** [comment]
@TomAugspurger thanks, didn't have much time today and thankfully this is such a small change so I just redid the changes. I did add one more test assertion this morning upon Jeff's recommendation, and that's included here as well.

**Comment by jreback** [comment]
thanks @RobinFiveWords

**Pull request closed without merging**